### ISSUE TITLE
`dappfile` format discussion: link to "contexts"/"environments" via ipfs?

### ISSUE THREAD

**Issue by nmushegian** [nexusdev/dapple#98]
Perhaps instead of specifying the constant map in the dappfile, it should point to an ipfs hash of the json/yaml object?

The motivation is that a "partial environment" could be used with a `step` but potentially live outside of a package: https://github.com/NexusDevelopment/dapple/issues/97

**Comment by ryepdx** [comment]
Not sure that specifying an IPFS hash instead of the map itself improves decoupling at all. I think it's better to have a human-readable map in the dappfile anyway. If you want to have ephemeral environment constants, then putting that information in another location makes sense.

Which then brings us back to the `--force` v.s. `--save` question. `--force` without `--save` makes sense if every environment is implicitly saved for distribution. `--force` and `--save` together make sense in an environment where Dapple keeps maps for distribution and maps for local development separate.

**Comment by nmushegian** [comment]
ok, so how about `--save` writes to an `env` dir which contains constant maps for different environments, while `--force` is for overwriting a constant which is under dapple versioning in the dappfile. It needs to be clear which is used when.

**Comment by ryepdx** [comment]
Yeah, that sounds good to me. Might also make sense to rename "dapple_packages" to "dapple/packages" and have it next to the "env" directory in "dapple/env".

**Comment by nmushegian** [comment]
Why not just `packages` and `env`, just like `contracts` and `build`?

**Comment by ryepdx** [comment]
That could work if we make them configurable like `contracts` and `build` are. The downside, IMO, is that there then is nothing to point people to Dapple if they come across the project in the wild with no foreknowledge of Dapple's default directory structure. Putting configuration data inside a `dapple` directory resolves that issue.

**Comment by nmushegian** [comment]
Perhaps there will be some kind of hint in the dappfile, like a `dapple_version` which may be useful later?

**Comment by ryepdx** [comment]
Yeah, that could work.

So this doesn't feel too much like polluting the user's project space to you? I'm starting to get that feeling, but then again I also argued for putting everything into a hidden directory.

**Comment by nmushegian** [comment]
In my projects which have frontends or off-chain components I have the dapple package in `dapp` or `packagename-dapp`. I think of dapple's scope as pretty narrow, maybe appropriate to use in part of a bigger build pipeline.

**Comment by dbrock** [comment]
I think `.dapple` would be a really good choice of directory name for miscellaneous "internal" state and configuration data.

**Comment by ryepdx** [comment]
I think we've gotten away from the original intent of this discussion, which was to determine how best to enable the use of environments resulting from partial execution of deploy scripts in conjunction with `dapple step`.

Environments don't seem large enough to me to warrant the extra level of indirection. They fit well into the YAML format the dappfile uses already. However, I definitely understand the motivation behind your suggestion. Maybe it would make sense to allow referencing external object maps from the CLI on a run-by-run basis.

In other words, something like:

```
dapple step --env morden --objects <ipfs hash or other locator>
```
